# [단계 10] 브루트 포스
#### 가장 간단한 알고리즘인, 모든 경우의 수를 검사하는 브루트 포스 알고리즘을 배워 봅시다.	

##### No. 2798 블랙잭 
> **문제**<br>
카지노에서 제일 인기 있는 게임 블랙잭의 규칙은 상당히 쉽다. 카드의 합이 21을 넘지 않는 한도 내에서, 카드의 합을 최대한 크게 만드는 게임이다. 블랙잭은 카지노마다 다양한 규정이 있다. 한국 최고의 블랙잭 고수 김정인은 새로운 블랙잭 규칙을 만들어 상근, 창영이와 게임하려고 한다. 김정인 버전의 블랙잭에서 각 카드에는 양의 정수가 쓰여 있다. 그 다음, 딜러는 N장의 카드를 모두 숫자가 보이도록 바닥에 놓는다. 그런 후에 딜러는 숫자 M을 크게 외친다.<br>
이제 플레이어는 제한된 시간 안에 N장의 카드 중에서 3장의 카드를 골라야 한다. 블랙잭 변형 게임이기 때문에, 플레이어가 고른 카드의 합은 M을 넘지 않으면서 M과 최대한 가깝게 만들어야 한다.<br>
N장의 카드에 써져 있는 숫자가 주어졌을 때, M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 구해 출력하시오.<br>
<br>
**입력**<br>
첫째 줄에 카드의 개수 N(3 ≤ N ≤ 100)과 M(10 ≤ M ≤ 300,000)이 주어진다. 둘째 줄에는 카드에 쓰여 있는 수가 주어지며, 이 값은 100,000을 넘지 않는 양의 정수이다. 합이 M을 넘지 않는 카드 3장을 찾을 수 있는 경우만 입력으로 주어진다.<br>
<br>
**출력**<br>
첫째 줄에 M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 출력한다.<br>

In [ ]:
from itertools import combinations

In [ ]:
#카드 개수 N
#부른 수 M
# 21이 넘지 않으면서 max값 찾기
#뽑은 수들끼리 합 구하기 --> sum
#넘지 않는 max 값 찾기

N, M = map(int, input().split())
card= list(map(int, input().split()))

select = list(combinations(card, 3))   #리스트에서 3개를 뽑는 법 (조합)
sum = []
for i in range(len(select)):
    cb = select[i]
    cb_sum = 0
    for j in range(3):
        cb_sum += cb[j] 
# print(cb_sum)
        if cb_sum <= M:
            sum.append(cb_sum)
print(sum)
print(max(sum))
#print(sum.index(500))
# 왜 예제 입력 2를 하면 500이 나올까요?

10 500
93 181 245 214 315 36 185 138 216 295
[93, 274, 93, 274, 488, 93, 274, 93, 274, 310, 93, 274, 459, 93, 274, 412, 93, 274, 490, 93, 274, 93, 338, 93, 338, 93, 338, 374, 93, 338, 93, 338, 476, 93, 338, 93, 338, 93, 307, 93, 307, 343, 93, 307, 492, 93, 307, 445, 93, 307, 93, 307, 93, 408, 444, 93, 408, 93, 408, 93, 408, 93, 408, 93, 129, 314, 93, 129, 267, 93, 129, 345, 93, 129, 424, 93, 278, 416, 93, 278, 494, 93, 278, 93, 231, 447, 93, 231, 93, 309, 181, 426, 181, 426, 181, 426, 462, 181, 426, 181, 426, 181, 426, 181, 426, 181, 395, 181, 395, 431, 181, 395, 181, 395, 181, 395, 181, 395, 181, 496, 181, 496, 181, 496, 181, 496, 181, 496, 181, 217, 402, 181, 217, 355, 181, 217, 433, 181, 217, 181, 366, 181, 366, 181, 366, 181, 319, 181, 319, 181, 397, 245, 459, 245, 459, 495, 245, 459, 245, 459, 245, 459, 245, 459, 245, 245, 245, 245, 245, 245, 281, 466, 245, 281, 419, 245, 281, 497, 245, 281, 245, 430, 245, 430, 245, 430, 245, 383, 245, 383, 245, 461, 214, 214, 214, 214, 214, 214, 

> 파이썬에서 조합을 구하는 방법을 처음 알게 됨
`from itertools import combinations` 

https://ourcstory.tistory.com/414
https://mildchae.tistory.com/5




##### No. 7568 덩치
> **문제**<br>
우리는 사람의 덩치를 키와 몸무게, 이 두 개의 값으로 표현하여 그 등수를 매겨보려고 한다. 어떤 사람의 몸무게가 x kg이고 키가 y cm라면 이 사람의 덩치는 (x, y)로 표시된다. 두 사람 A 와 B의 덩치가 각각 (x, y), (p, q)라고 할 때 x > p 그리고 y > q 이라면 우리는 A의 덩치가 B의 덩치보다 "더 크다"고 말한다. 예를 들어 어떤 A, B 두 사람의 덩치가 각각 (56, 177), (45, 165) 라고 한다면 A의 덩치가 B보다 큰 셈이 된다. 그런데 서로 다른 덩치끼리 크기를 정할 수 없는 경우도 있다. 예를 들어 두 사람 C와 D의 덩치가 각각 (45, 181), (55, 173)이라면 몸무게는 D가 C보다 더 무겁고, 키는 C가 더 크므로, "덩치"로만 볼 때 C와 D는 누구도 상대방보다 더 크다고 말할 수 없다.<br>
N명의 집단에서 각 사람의 덩치 등수는 자신보다 더 "큰 덩치"의 사람의 수로 정해진다. 만일 자신보다 더 큰 덩치의 사람이 k명이라면 그 사람의 덩치 등수는 k+1이 된다. 이렇게 등수를 결정하면 같은 덩치 등수를 가진 사람은 여러 명도 가능하다. 아래는 5명으로 이루어진 집단에서 각 사람의 덩치와 그 등수가 표시된 표이다.<br>
이름	(몸무게, 키)	덩치 등수<br>
A	(55, 185)	2<br>
B	(58, 183)	2<br>
C	(88, 186)	1<br>
D	(60, 175)	2<br>
E	(46, 155)	5<br>
위 표에서 C보다 더 큰 덩치의 사람이 없으므로 C는 1등이 된다. 그리고 A, B, D 각각의 덩치보다 큰 사람은 C뿐이므로 이들은 모두 2등이 된다. 그리고 E보다 큰 덩치는 A, B, C, D 이렇게 4명이므로 E의 덩치는 5등이 된다. 위 경우에 3등과 4등은 존재하지 않는다. 여러분은 학생 N명의 몸무게와 키가 담긴 입력을 읽어서 각 사람의 덩치 등수를 계산하여 출력해야 한다.<br>
<br>
**입력**<br>
첫 줄에는 전체 사람의 수 N이 주어진다. 그리고 이어지는 N개의 줄에는 각 사람의 몸무게와 키를 나타내는 양의 정수 x와 y가 하나의 공백을 두고 각각 나타난다.<br>
<br>
**출력**<br>
여러분은 입력에 나열된 사람의 덩치 등수를 구해서 그 순서대로 첫 줄에 출력해야 한다. 단, 각 덩치 등수는 공백문자로 분리되어야 한다.

In [2]:
N = int(input())
h = []

for i in range(N):
    x, y = map(int, input().split())
    h.append((x, y))

for a in h :
    rank = 1
    for b in h:
        if (a[0] < b[0]) and (a[1] < b[1]):
            rank += 1
    print(rank, end = ' ')

5
55 185
58 183
88 186
60 175
46 155
2 2 1 2 5 

In [1]:
x kg, y cm   (x, y)
A (x, y)
B (p, q)
x > p and y > q   A덩치 B덩치 보다 크다
N명 집단
덩치 등수: 자신보다 더 "큰 덩치"의 사람의 수 > , k명 있으면 등수 = k+1
https://roseline124.github.io/algorithm/2019/04/06/Altorithm-baekjoon-7568.html
https://claude-u.tistory.com/122

SyntaxError: ignored